### PR TITLE
SDK: scoping styles with CSS-modules using Babel

### PR DESCRIPTION
**Pull request by simison**
Adds support for [CSS modules](https://github.com/css-modules/css-modules) mainly to scope CSS when building extensions with the SDK — relying to existing `className="selector"` -syntax.

Read more: pa9srA-E-p2

Uses [`babel-plugin-react-css-modules`](https://github.com/gajus/babel-plugin-react-css-modules) to resolve plain CSS classes to CSS modules.

### Example

**Normally using CSS modules looks like this:**
```js
import styles from './styles.scss';

<Example className={ styles.selectorName } />
```
(see previous [bare-bones exploration](https://github.com/Automattic/wp-calypso/pull/26933) how that would look like in Calypso)

**With this PR, however, we keep doing what we've always done:**
```js
import './styles.scss';

<Example className='selector-name' />
// or
const example = 'selector-name';
<Example className={ example } />
```

**Output is scoped by hashing:**
```
.3-F28 { }
```
```html
<div class="3-F28"></div>
```

Normally, `babel-plugin-react-css-modules` requires you to use `styleName` attribute in JSX instead of `className`. Since we don't want to change everything in Calypso, we rename all the `className` attributes with Babel. That is; during the build, `className` becomes `styleName` and is thus picked up by `babel-plugin-react-css-modules` and the result is using CSS-modules "under the hood" and compiled CSS is scoped. 👌 

### Pros
- The output is very compact and well scoped

- No changes required to Calypso components

- ~Unlike bare-bones CSS modules, does _not_ require using camel case CSS class-names~ ([I stand corrected!](https://github.com/Automattic/wp-calypso/pull/26976#issuecomment-417645033))

- Allows using multiple CSS class-names just like you'd normally do with `className`

- Supports string literals (`className='lorem'`) and expressions (`className={ lorem }`)
   - When the value cannot be determined at the compile time, the plugin inlines all possible styles into the file. It then uses `getClassName` helper function to resolve the value at the runtime.

### Cons
(Quite a few but perhaps they can be worked out...)

- For a developer, it might not be obvious that this build step is going to happen for their code

- No global selectors; all used CSS class-names have to be imported in the file. This is a good thing since we want to enforce declaring CSS dependencies. Without our custom `className`→`styleName` transform, global selectors would be set via regular `className` attribute. We _could_ introduce a new attribute (e.g. `globalClassName`) if needed. I'm not sure how this affects e.g. `body.rtl` and other global selectors like that; we might need to turn those into attribute selectors instead — and that's a good thing: `.rtl` becomes `[dir='rtl']`

- Unresolved class-names (i.e. globals) throw at build time for string literals but for expressions, they throw only at run-time. This introduces some shakiness that we could mitigate by setting the plugin to only _warn_ in production mode instead of throwing an error:
    ```js
    const problems = 'does-not-exist';
    <Example className={ problems } />

    // → Error: Could not resolve the styleName 'does-not-exist'.
    ``` 
    I bet we have a bunch of un-used class-names littered around Calypso so this would help to clean them out. 

- When importing multiple stylesheets, we must use [a named reference](https://github.com/gajus/babel-plugin-react-css-modules/tree/f635b3570e7da27dc71191f65ad4c6654b81f900#named-stylename-resolution).

    That is, this won't work:
    ```js
    import './styles1.scss'; // has lorem
    import './styles2.scss'; // has ipsum

    <Example className='lorem ipsum' />
    ```

    Nor will this:
    ```js
    import './styles1.scss'; // inside styles1.scss: @import './styles2.scss';

    <Example className='lorem ipsum' />
    ```

    Only this works:
    ```js
    import styles1 from './styles1.scss';
    import styles2 from './styles2.scss';

    <Example className='styles1.lorem styles2.ipsum' />
    ```

    This is great for clarity and following single responsibility principle but I bet our CSS is full of examples that would require some changes to follow this model. Since we don't import _any_ stylesheets in Calypso right now, we would not be forced to do any visible changes at this time. However, it is a limitation developer might not expect and since the build step is a bit of black-box magic, it might not be clear _why_ they need to do this.

-  Does not scope element CSS selectors like `body` or `p` — we need to deal with those separately. We're not importing anything like that in Gutenberg block styles currently so we're good for now. Nothing breaks when using element selectors with this PR so it's a separate concern for the SDK.

- Makes it impossible to overwrite styles for Calypso components in wp-admin

### Testing

Compile a block:
```
npm run sdk:gutenberg -- --editor-script=client/gutenberg/extensions/editor-notes/index.js
```

Test the block in Gutenberg, you should see:

screen shot 2018-08-28 at 10 14 46

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -101,22 +101,30 @@ const wordpressExternals = ( context, request, callback ) =>
...
-		cssFilename ||
-		( isDevelopment || calypsoEnv === 'desktop' ? '[name].css' : '[name].[chunkhash].css' );
+function getWebpackConfig(
+	{ cssModules = false, cssFilename, externalizeWordPressPackages = false } = {},
+	// eslint-disable-next-line no-unused-vars
+	argv
+) {
+	// Temporary solution
+	// Used in `babel.config.js`
+	cssModules && ( process.env.CSS_MODULES = 'true' );
```
Comment (simison) [inline webpack.config.js]: This is here to pass config to `babel.config.js` — I'll look for a better way to do this.

**Comment by ockham** [comment]
> Unlike bare-bones CSS modules, does _not_ require using camel case CSS class-names

Just a quick note, I think that wouldn't be a hard requirement:

```js
import styles from style.scss;

export default const MyComponent = () => (
	<div className={ styles[ 'my-component__some-style' ] }>
		No style, no grace
	</div>
);

**Comment by ockham** [comment]
> - Supports string literals (`className='lorem'`) and expressions (`className={ lorem }`)
>   - When the value cannot be determined at the compile time, the plugin inlines all possible styles into the file. It then uses `getClassName` helper function to resolve the value at the runtime.

We use `classNames` quite heavily to set `className` conditionally, e.g. depending on a given props value. Would this mean that we'd bloat files that do something like that a lot -- more so than with other conceivable solutions?

**Comment by simison** [comment]
> > Unlike bare-bones CSS modules, does not require using camel case CSS class-names
>
>Just a quick note, I think that wouldn't be a hard requirement:

Right on! I stand corrected.

File: client/gutenberg/extensions/editor-notes/style.scss [hunk 2]
```diff
@@ -1,17 +1,4 @@
...
-		padding-top: 6px;
-	}
-}
-
-.editor-notes__editor-indicator {
-	position: absolute;
-	top: 0;
-	left: 8px;
-	font-style: italic;
+.wp-block-a8c-editor-notes .card {
```
Comment (sirreal) [inline client/gutenberg/extensions/editor-notes/style.scss]: Is this here to ensure `card` is correctly overridden?

See https://github.com/Automattic/wp-calypso/pull/26933#discussion_r214826802

File: babel.config.js [hunk 3]
```diff
@@ -55,6 +69,31 @@ const config = {
...
+				to: 'styleName',
+			},
+		],
+		cssModules && [
+			'babel-plugin-react-css-modules',
+			{
+				context: __dirname,
+				filetypes: {
+					'.scss': {
+						syntax: 'postcss-scss',
```
Comment (sirreal) [inline babel.config.js]: Would the goal here be to drop our independent SASS compilers? Is https://github.com/postcss/postcss-scss sufficient? It doesn't sound like a "full" sass compiler.

**Review by sirreal — COMMENTED** [review]
I'm not sure of the best way to test the compiled block in Gutenberg.

**Pull request closed without merging**